src/sentry/notifications/platform/providers/base.py

```diff
@@ -0,0 +1,44 @@
+import abc
+from enum import StrEnum
+from typing import Generic, TypeVar
+
+from sentry.organizations.services.organization.model import RpcOrganizationSummary
+
+ProviderRenderable = TypeVar("ProviderRenderable")
+"""
+A renderable object that is understood by the notification provider.
+For example, Email might expect HTML, or raw text; Slack might expect a JSON Block Kit object.
+"""
+
+
+class NotificationProviderKey(StrEnum):
+    """
+    The unique keys for each registered notification provider.
+    """
+
+    EMAIL = "email"
+    SLACK = "slack"
+    MSTEAMS = "msteams"
+    DISCORD = "discord"
+
+
+class NotificationProvider(abc.ABC, Generic[ProviderRenderable]):
+    """
+    A base class for all notification providers.
+    """
+
+    @property
+    @abc.abstractmethod
+    def key(self) -> NotificationProviderKey:
+        """
+        The `NotificationProviderKey` associated with this provider.
+        """
+        raise NotImplementedError("Must use a NotificationProviderKey")
+
+    @classmethod
+    @abc.abstractmethod
+    def is_available(cls, *, organization: RpcOrganizationSummary | None = None) -> bool:
+        """
+        Returns `True` if the provider is available given the key word arguments.
```

Comment:
Is this for conditionally enabling providers in some way? I like the concept, but am curious if we need to tie this into other systems like ACI's workflow action selectors.

Reply:
It'd make sense, we had a similar concept for notification actions but essentially a way to communicate to other sentry services which providers are available to connect with given context:

https://github.com/getsentry/sentry/blob/master/src/sentry/notifications/api/endpoints/notification_actions_available.py#L32-L38

In this case the only context is currently organization, but maybe notification type, or something can be passed in the future to determine the available providers.

Reply:
> if we need to tie this into other systems like ACI's workflow action selectors. 
As for this point, I'd rather the notification platform own it's own availability independently and we formalize an API or access point for all other services since I can picture a world where other stuff than ACI will want to know what notifs an org has access to. For now that is `provider_registry.get_available()` but we can revisit that, i'll add a mention in Linear